File: examples/service/ExampleService.java
package examples.service;

import com.roslib.ros.CallbackSrvT;
import com.roslib.ros.Msg;
import com.roslib.ros.NodeHandle;
import com.roslib.ros.ServiceServer;
import com.roslib.tinyros_hello.Test;

public class ExampleService {

    public static void main(String[] args) throws InterruptedException {
        NodeHandle nh = new NodeHandle();
        while (!nh.initNode("127.0.0.1")) {
            System.out.println("Java: initNode failed.");
            Thread.sleep(500);
        }

        nh.advertiseService(new ServiceServer<Test.TestRequest, Test.TestResponse>("test_srv", new CallbackSrvT() {
            @Override
            public void callback(Msg req, Msg res) {
                ((Test.TestResponse)res).output = "Hello, tiny-ros ^_^";
            }
        }, new Test.TestRequest(), new Test.TestResponse()));

        while(nh.ok()) {
            nh.spinOnce();
            Thread.sleep(500);
        }

    }

}

File: examples/service/ExampleService.go
package main

import (
    "time"
    "tiny_ros/tinyros"
    "tiny_ros/tinyros_hello"
)

func service_cb(req tinyros.Msg, resp tinyros.Msg) {
    tresp := resp.(*tinyros_hello.TestResponse)
    tresp.Go_output = "Hello, tiny-ros ^_^"
}

func main() {
    tinyros.Go_init("GoExampleService", "127.0.0.1")
    
    server := tinyros.NewServiceServer("test_srv", service_cb, tinyros_hello.NewTestRequest(), tinyros_hello.NewTestResponse())

    tinyros.Go_nh().Go_advertiseService(server)
    
    for {
        time.Sleep(10 * time.Second)
    }
}
